zallet/src/components/json_rpc/methods/get_address_for_account.rs

```diff
@@ -0,0 +1,208 @@
+            None => Err(LegacyCode::Wallet.with_message(format!(
+                "Error: diversifier index {} cannot generate an address with the requested receivers.",
+                u128::from(diversifier_index),
+            ))),
+        }
+    } else {
+        wallet
+            .get_next_available_address(account_id, request)
+            .map_err(|e| map_sqlite_error(e, &account))?
+            // This would be a race condition between this and account deletion.
+            .ok_or(RpcErrorCode::InternalError.into())
+    }?;
+
+    // Only include `account` in the response if it was provided in the request. We rely
+    // on the checks performed in `parse_account_parameter`.
+    let account = match account {
+        JsonValue::Number(n) => n.as_u64(),
+        _ => None,
+    };
+
+    Ok(Address {
+        uuid: account_id.expose_uuid().to_string(),
+        account,
+        diversifier_index: diversifier_index.into(),
+        receiver_types,
+        address: address.encode(wallet.params()),
+    })
+}
+
+fn map_sqlite_error(e: SqliteClientError, account: &JsonValue) -> ErrorObjectOwned {
+    match e {
+        SqliteClientError::TransparentDerivation(error) => LegacyCode::Wallet.with_message(format!(
+            "Error: failed to derive a transparent component: {}",
+            error,
+        )),
+        SqliteClientError::AddressGeneration(e) => match e {
+            AddressGenerationError::InvalidTransparentChildIndex(diversifier_index) => {
+                LegacyCode::Wallet.with_message(format!(
+                    "Error: diversifier index {} cannot generate an address with a transparent receiver.",
+                    u128::from(diversifier_index),
+                ))
+            }
+            AddressGenerationError::InvalidSaplingDiversifierIndex(diversifier_index) => {
+                LegacyCode::Wallet.with_message(format!(
+                    "Error: diversifier index {} cannot generate an address with a Sapling receiver.",
+                    u128::from(diversifier_index),
+                ))
+            }
+            AddressGenerationError::DiversifierSpaceExhausted => LegacyCode::Wallet.with_static(
+                "Error: ran out of diversifier indices. Generate a new account with z_getnewaccount"
+            ),
+            AddressGenerationError::ReceiverTypeNotSupported(typecode) => match typecode {
+                unified::Typecode::P2sh =>  LegacyCode::Wallet.with_static(
+                    "Error: P2SH addresses can not be created using this RPC method.",
+                ),
+                _ => LegacyCode::Wallet.with_message(format!(
+                    "Error: receiver type {:?} is not supported.",
+                    typecode,
+                ))
+            },
+            AddressGenerationError::KeyNotAvailable(typecode) => LegacyCode::Wallet.with_message(format!(
+                "Error: account {} does not have a receiver component with type {:?}.",
```

Comment:
```suggestion
                "Error: account {} does not have a key item with type {:?}.",
```

Reply:
Going with "cannot generate a receiver component" instead as that is clearer to the user (the RPC docs talk about receiver types, not key items).